--- a/ext/socket/extconf.rb
+++ b/ext/socket/extconf.rb
@@ -618,65 +618,73 @@
     "udpsocket.#{$OBJEXT}",
     "unixsocket.#{$OBJEXT}",
     "unixserver.#{$OBJEXT}",
     "option.#{$OBJEXT}",
     "ancdata.#{$OBJEXT}",
     "raddrinfo.#{$OBJEXT}",
     "ifaddr.#{$OBJEXT}"
   ]
 
   if getaddr_info_ok == :wide
     if !have_type("struct in6_addr", headers) and have_type("struct in_addr6", headers)
       $defs.pop(2)
       $defs << "-Din_addr6=in6_addr"
     end
     if have_struct_member("struct in6_addr", "s6_addr8", headers)
       $defs[-1] = "-Ds6_addr=s6_addr8"
     end
     if ipv6 == "kame" && have_struct_member("struct in6_addr", "s6_addr32", headers)
       $defs[-1] = "-DFAITH"
     end
     $CPPFLAGS="-I. "+$CPPFLAGS
     $objs += ["getaddrinfo.#{$OBJEXT}"]
     $objs += ["getnameinfo.#{$OBJEXT}"]
     $defs << "-DGETADDRINFO_EMU"
   end
 
   # workaround for recent Windows SDK
   $defs << "-DIPPROTO_IPV6=IPPROTO_IPV6" if $defs.include?("-DHAVE_CONST_IPPROTO_IPV6") && !have_macro("IPPROTO_IPV6")
 
   $distcleanfiles << "constants.h" << "constdefs.*"
 
   if enable_config("socks", ENV["SOCKS_SERVER"])
     if have_library("socks5", "SOCKSinit")
       $defs << "-DSOCKS5" << "-DSOCKS"
     elsif have_library("socksd", "Rconnect") || have_library("socks", "Rconnect")
       $defs << "-DSOCKS"
     end
   end
 
   hdr = "netinet6/in6.h"
-  if /darwin/ =~ RUBY_PLATFORM and !try_compile(<<"SRC", nil, :werror=>true)
+  /darwin/ =~ RUBY_PLATFORM and
+  checking_for("if apple's #{hdr} needs s6_addr patch") {!try_compile(<<"SRC", nil, :werror=>true)} and
 #include <netinet/in.h>
 int t(struct in6_addr *addr) {return IN6_IS_ADDR_UNSPECIFIED(addr);}
 SRC
-    print "fixing apple's netinet6/in6.h ..."; $stdout.flush
-    in6 = File.read("/usr/include/#{hdr}")
+  checking_for("fixing apple's #{hdr}", "%s") do
+    file = xpopen(%w"clang -include netinet/in.h -E -xc -", in: IO::NULL) do |f|
+      re = %r[^# *\d+ *"(.*/netinet/in\.h)"]
+      Logging.message "  grep(#{re})\n"
+      f.read[re, 1]
+    end
+    Logging.message "Substitute from #{file}\n"
+
+    in6 = File.read(file)
     if in6.gsub!(/\*\(const\s+__uint32_t\s+\*\)\(const\s+void\s+\*\)\(&(\(\w+\))->s6_addr\[(\d+)\]\)/) do
         i, r = $2.to_i.divmod(4)
         if r.zero?
           "#$1->__u6_addr.__u6_addr32[#{i}]"
         else
           $&
         end
       end
       FileUtils.mkdir_p(File.dirname(hdr))
-      open(hdr, "w") {|f| f.write(in6)}
+      File.write(hdr, in6)
       $distcleanfiles << hdr
       $distcleandirs << File.dirname(hdr)
-      puts "done"
+      "done"
     else
-      puts "not needed"
+      "not needed"
     end
   end
   create_makefile("socket")
 end
